tests/ci/pr_info.py

```diff
@@ -304,57 +304,66 @@ def has_changes_in_submodules(self):
-            if (
-                not f.startswith("tests/queries")
-                or not f.startswith("tests/integration")
-                or not f.startswith("tests/performance")
-            ):
-                return False
+        return not any(
+            f.startswith("programs")
+            or f.startswith("src")
+            or f.startswith("base")
+            or f.startswith("cmake")
+            or f.startswith("rust")
+            or f == "CMakeLists.txt"
+            or f == "tests/ci/build_check.py"
+            for f in self.changed_files
+        )
 
-        return True
 
-    def can_skip_integration_tests(self):
-        # TODO: See a broken loop
-        if FORCE_TESTS_LABEL in self.labels:
-            return False
+def can_skip_integration_tests(self, versions: List[str]):
+    if FORCE_TESTS_LABEL in self.labels:
+        return False
 
-        if self.changed_files is None or not self.changed_files:
-            return False
+    # If docker image(s) relevant to integration tests are updated
+    if self.sha not in versions:
+        return False
 
-        for f in self.changed_files:
-            # TODO: this logic is broken, should be fixed before using
-            if not f.startswith("tests/queries") or not f.startswith(
-                "tests/performance"
-            ):
-                return False
+    if self.changed_files is None or not self.changed_files:
+        return False
 
-        return True
+    if not self.can_skip_builds_and_use_version_from_master():
+        return False
 
-    def can_skip_functional_tests(self):
-        # TODO: See a broken loop
-        if FORCE_TESTS_LABEL in self.labels:
-            return False
+    # Integration tests can be skipped if integration tests are not changed
+    return not any(
+        f.startswith("tests/integration/") or f == "tests/ci/integration_test_check.py"
+        for f in self.changed_files
+    )
 
-        if self.changed_files is None or not self.changed_files:
-            return False
 
-        for f in self.changed_files:
-            # TODO: this logic is broken, should be fixed before using
-            if not f.startswith("tests/integration") or not f.startswith(
-                "tests/performance"
-            ):
-                return False
+def can_skip_functional_tests(self, version):
+    if FORCE_TESTS_LABEL in self.labels:
+        return False
+
+    # If docker image(s) relevant to functional tests are updated
+    if self.sha not in version:
+        return False
+
+    if self.changed_files is None or not self.changed_files:
+        return False
 
-        return True
+    if not self.can_skip_builds_and_use_version_from_master():
+        return False
+
+    # Functional tests can be skipped if queries tests are not changed
+    return not any(
+        f.startswith("tests/queries/0_stateless")
+        or f.startswith("tests/queries/1_stateful")
```

Comment:
yeah.. we still need to somehow identify if we run stateless or stateful checks.

Maybe we can get it from the docker image name. It either `clickhouse/stateless-test` or `clickhouse/stateful-test`, so we could pass `test_type: Literal["stateless", "stateful"]` as the method argument from the main function